### PR TITLE
Tackling some low hanging performance fruit

### PR DESCRIPTION
**Pull request by davidfowl**
- Use native Memory/Span APIs on Stream in .NET Core 2.0
- Remove double copying in formatters
- Implemented custom HttpContent over ReadOnlyBuffer<byte>

### CODE REVIEW

File: src/Common/WebSocketExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+{
+    internal static class WebSocketExtensions
+    {
+        public static async Task SendAsync(this WebSocket webSocket, ReadOnlyBuffer<byte> buffer, WebSocketMessageType webSocketMessageType, CancellationToken cancellationToken = default)
+        {
+            // TODO: Consider chunking writes here if we get a multi segment buffer
+#if NETCOREAPP2_1
+            if (buffer.IsSingleSegment)
+            {
+                await webSocket.SendAsync(buffer.First, webSocketMessageType, endOfMessage: true, CancellationToken.None);
```
Comment (BrennanConroy) [inline src/Common/WebSocketExtensions.cs]: Pass in `cancellationToken`

File: src/Microsoft.AspNetCore.SignalR.Common/Internal/Formatters/BinaryMessageFormatter.cs [hunk 2]
```diff
@@ -2,21 +2,23 @@
...
+        public static void WriteLengthPrefix(long length, Stream output)
         {
             // This code writes length prefix of the message as a VarInt. Read the comment in
             // the BinaryMessageParser.TryParseMessage for details.
 
-            var lenBuffer = stackalloc byte[5];
+#if NETCOREAPP2_1
+            Span<byte> lenBuffer = stackalloc byte[5];
+#else
+            var lenBuffer = new byte[5];
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.SignalR.Common/Internal/Formatters/BinaryMessageFormatter.cs]: Why isn't this stackalloc?

File: src/Common/StreamExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
+﻿using System.Buffers;
```
Comment (BrennanConroy) [inline src/Common/StreamExtensions.cs]: Missed a few copyrights

File: src/Microsoft.AspNetCore.SignalR.Common/Internal/Formatters/BinaryMessageFormatter.cs (hunk 2 above)
Comment (benaadams) [reply src/Microsoft.AspNetCore.SignalR.Common/Internal/Formatters/BinaryMessageFormatter.cs]: Unsafe?

File: src/Common/StreamExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+
+namespace System.IO
+{
+    internal static class StreamExtensions
+    {
+        public static async Task WriteAsync(this Stream stream, ReadOnlyBuffer<byte> buffer)
+        {
+            // REVIEW: Should we special case IsSingleSegment here?
```
Comment (anurse) [inline src/Common/StreamExtensions.cs]: To remove the state machine? Let's wait until we see it show up in perf data.

File: src/Common/WebSocketExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.WebSockets
+{
+    internal static class WebSocketExtensions
+    {
+        public static async Task SendAsync(this WebSocket webSocket, ReadOnlyBuffer<byte> buffer, WebSocketMessageType webSocketMessageType, CancellationToken cancellationToken = default)
```
Comment (anurse) [inline src/Common/WebSocketExtensions.cs]: Looks like these are all tail-`awaits`, no need for `async`?

File: src/Microsoft.AspNetCore.SignalR.Common/Internal/Formatters/BinaryMessageFormatter.cs (hunk 2 above)
Comment (anurse) [reply src/Microsoft.AspNetCore.SignalR.Common/Internal/Formatters/BinaryMessageFormatter.cs]: Also, does `Stream` have a `WriteAsync` that takes `byte*`? We used to copy to a pooled array before, which seems unnecessary.

**Pull request merged**